src/libtorch.cc

```diff
@@ -740,8 +740,11 @@ ModelInstanceState::ValidateInputs(const size_t expected_input_cnt)
       // input names since they are the keys for the dictionary
       input_index_map_[io_name] = i;
     } else {
-      // input tensor name must be in 'allowed_inputs' or must follow the naming
-      // convention
+      // Rules for (non-Dictionary) input tensor names:
+      // 1. Must be in 'allowed_inputs' (arguments in the forward function)
+      // 2. Must follow the naming convention i.e. <name>__<index>
+      // 3. If neither of the above conditions are satisfied, enforce strict
+      // ordering of model inputs.
```

Comment:
Shouldn't you check and decide which of the rules should be applied for all inputs (and outputs below) first? What happens if user provides a config with 3. in mind but the name happens to match one of the `allowed_inputs`

Reply:
I agree with fixing the edge case here. 
@CoderHam But can you remind me the reason behind this change? As in why  3 is needed if user can enforce ordering using the naming convention?

Reply:
> Shouldn't you check and decide which of the rules should be applied for all inputs (and outputs below) first? What happens if user provides a config with 3. in mind but the name happens to match one of the allowed_inputs

A user may want to use a mixture of the naming convention i.e. use input names from the model's forward function but random output names from the output. 

I agree that we can enforce that all the inputs (and outputs) follow the same naming convention. That being said we do not check for a collision. For example if the user specifies INPUTA__0 and INPUTB__0 the model would be loaded correctly but see issues during execution. Are you suggesting we add tests for such collisions?

Reply:
> But can you remind me the reason behind this change? As in why 3 is needed if user can enforce ordering using the naming convention?

The reason for adding the third case is for users who want to specify arbitrary names for I/O. They need only know the number of Inputs and Outputs and not their names.

Reply:
I think the ask here is how do you plan to handle cases where the model forward function has the following inputs in order:
1. INPUT0, INPUT1, INPUT2, INPUT3
But the input in config are following :

2. INPUT4, INPUT5, INPUT0, INPUT1

Reply:
That would work without my new changes for cases where the inputs are already in the forward function, the order is extracted from the allowed_inputs i.e. the order in which they are present in the model's forward function.

Reply:
INPUT4 and INPUT5 are not in the forward function right?

May be the mapping is(using 3): INPUT0->INPUT4, INPUT1->INPUT5, INPUT2->INPUT0, INPUT3->INPUT1
Or may be the mapping is(using partially 1): INPUT0->INPUT0, INPUT1-INPUT1, INPUT2->INPUT4, INPUT3->INPUT5

Reply:
@tanmayv25 I have enforced the need for consistency of input / output naming conventions to avoid this issue. I have not added logic to detect collisions yet. For inputs there can be multi-way collisions due to the 3 choices for naming format.  

please re-review